libtiny_client/src/state.rs

```diff
@@ -512,6 +671,50 @@ impl StateInner {
             }
         }
     }
+
+    /// If channel is in Joining state cancel Joining task, otherwise sent part message
+    fn leave_channel(&mut self, msg_chan: &mut Sender<Cmd>, chan: &str) {
+        if let Some(idx) = utils::find_idx(&self.chans, |c| c.name == chan) {
+            match &mut self.chans[idx].join_state {
+                JoinState::NotJoined => {
+                    warn!("Unexpected state. Leaving channel before attempting to join.")
+                }
+                JoinState::Joining { stop_task, .. } => {
+                    debug!("Aborting task to retry joining {}", chan);
+                    stop_task.try_send(()).unwrap();
+                }
+                JoinState::Joined => msg_chan.try_send(Cmd::Msg(wire::part(chan))).unwrap(),
+            }
+        }
+    }
+
+    /// Kills all tasks that are trying to join channels
+    fn kill_join_tasks(&mut self) {
+        for chan in &mut self.chans {
+            if let JoinState::Joining { stop_task } = &mut chan.join_state {
+                stop_task.try_send(()).unwrap();
```

Comment:
```suggestion
                let _ = stop_task.try_send(());
```

When we call this there will sometimes be a race between the `retry_channel_join` task and this code. For example it's possible that `retry_channel_join` just sent a join message and returned (so the receiving end of the channel is closed) but we haven't heard back from the server yet so we didn't update `JoinState` yet and the sending end is still available. This is a benign race that we can ignore.